syntax = "proto3";

import "google/protobuf/struct.proto";

service Greeter {
  rpc SayHello (HelloRequest) returns (HelloReply) {}
  rpc SayHelloManyTimes (HelloRequest) returns (stream HelloReply) {}
  rpc SayHelloClientStream (stream HelloRequest) returns (HelloReply) {}
  rpc SayHelloBidirectional (stream HelloRequest) returns (stream HelloReply) {}
  rpc SayHelloWithStruct (google.protobuf.Struct) returns (HelloReply) {}
  rpc SayComplexHello (ComplexRequest) returns (HelloReply) {}
  rpc SayComplexHelloStream (stream ComplexRequest) returns (HelloReply) {}
}

message HelloRequest {
  string name = 1;
}

message HelloReply {
  string message = 1;
}

message ComplexRequest {
  string name = 1;
  google.protobuf.Struct data = 2;
}